In [47]:
#!pip uninstall tensorflow
#!pip install tensorflow==2.17.1
#!pip install protobuf==3.20.3

In [48]:
import numpy as np
import tensorflow as tf


In [49]:
print(np.__version__)

1.26.4


In [50]:
print(tf.__version__)

2.17.1


In [51]:
SEED = 42
np.random.seed(SEED)
tf.random.set_seed(SEED)

In [52]:
from tensorflow import keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense
from tensorflow.keras.optimizers import SGD

In [53]:
# Initialize the model
model = Sequential()

# Adding the convolutional layer Conv2D
model.add(Conv2D(filters=32, kernel_size=(3,3), activation = "relu", input_shape=(200,200,3)))

# Reduce the size of the featumire map with max pooling MaxPooling2D
model.add(MaxPooling2D(pool_size=(2, 2)))

# Turn the multi-dimensional result into vectors using a Flatten layer
model.add(Flatten())

# Add a dense layer with 64 neurons and 'relu' activation
model.add(Dense(64, activation='relu'))

# create the Dense layer with 1 neuron
model.add(Dense(1, activation='sigmoid'))

# Compile the model with SGD optimizer
optimizer = SGD(learning_rate=0.002, momentum=0.8) 
model.compile(optimizer=optimizer, loss='binary_crossentropy', metrics=['accuracy'])


#### Question 1 : 
Among the list and for the case of binary classification problem, the best loss function is binary crossentropy.

#### Question 2 : 

In [54]:
model.summary()

Model: "sequential_5"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_4 (Conv2D)               │ (None, 198, 198, 32)   │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_4 (MaxPooling2D)  │ (None, 99, 99, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_4 (Flatten)             │ (None, 313632)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_8 (Dense)                 │ (None, 64)             │    20,072,512 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_9 (Dense)                 │ (None, 1)              │            65 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 20,073,473 (76.57 MB)

 Trainable params: 20,073,473 (76.57 MB)

 Non-trainable params: 0 (0.00 B)

The total number of parameters of the model is 20,073,473. The close answer is 20 072 512. 

### Generators and Training

In [55]:
from tensorflow.keras.preprocessing.image import ImageDataGenerator

In [56]:
train_generator = ImageDataGenerator(rescale=1./255)

train_ds = train_generator.flow_from_directory(
    './data/train',
    target_size=(200, 200),  
    batch_size=20,
    class_mode='binary', 
    shuffle=True           
)

test_generator = ImageDataGenerator(rescale=1./255)

test_ds = test_generator.flow_from_directory(
    './data/test',
    target_size=(200, 200),  
    batch_size=20,
    class_mode='binary', 
    shuffle=True           
)

Found 800 images belonging to 2 classes.
Found 201 images belonging to 2 classes.


In [57]:
history = model.fit(train_ds, epochs=10, validation_data=test_ds)

Epoch 1/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 11s 251ms/step - accuracy: 0.5649 - loss: 0.6871 - val_accuracy: 0.6318 - val_loss: 0.6238
Epoch 2/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 10s 237ms/step - accuracy: 0.6991 - loss: 0.5857 - val_accuracy: 0.6368 - val_loss: 0.6439
Epoch 3/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 10s 237ms/step - accuracy: 0.7037 - loss: 0.5550 - val_accuracy: 0.6269 - val_loss: 0.6232
Epoch 4/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 10s 239ms/step - accuracy: 0.7298 - loss: 0.5632 - val_accuracy: 0.6567 - val_loss: 0.6155
Epoch 5/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 10s 237ms/step - accuracy: 0.7303 - loss: 0.5278 - val_accuracy: 0.6617 - val_loss: 0.5871
Epoch 6/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 9s 230ms/step - accuracy: 0.7620 - loss: 0.4704 - val_accuracy: 0.6617 - val_loss: 0.6067
Epoch 7/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 10s 238ms/step - accuracy: 0.7745 - loss: 0.4756 - val_accuracy: 0.6667 - val_loss: 0.5717
Epoch 8/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 9s 237ms/step - accuracy: 0.7895 - loss: 0.4494 - val_accura

In [58]:
history.history['accuracy']

[0.5987499952316284,
 0.6825000047683716,
 0.7024999856948853,
 0.7275000214576721,
 0.7350000143051147,
 0.7487499713897705,
 0.78125,
 0.7975000143051147,
 0.768750011920929,
 0.8262500166893005]

In [61]:
np.median(history.history['accuracy'])

0.7418749928474426

#### Question 3 :
Median of training accuracy for all the epochs for the model is 0.74, close to the answer of 0.72

In [62]:
np.std(history.history['loss'])

0.07758409520271932

#### Question 4 :
Standard deviation of training loss for all the epochs for the model is 0.077, close to the answer of 0.068

### Training the model using data augmentations

In [63]:
train_generator = ImageDataGenerator(
    rescale=1./255,
    rotation_range=50,
    width_shift_range=0.1,
    height_shift_range=0.1,
    zoom_range=0.1,
    horizontal_flip=True,
    fill_mode='nearest'
)

train_ds = train_generator.flow_from_directory(
    './data/train',
    target_size=(200, 200),  
    batch_size=20,
    class_mode='binary', 
    shuffle=True           
)

Found 800 images belonging to 2 classes.


In [64]:
history = model.fit(
    train_ds, 
    epochs=10, 
    validation_data=test_ds)

Epoch 1/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 15s 382ms/step - accuracy: 0.6380 - loss: 0.6672 - val_accuracy: 0.6766 - val_loss: 0.6423
Epoch 2/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 15s 368ms/step - accuracy: 0.6652 - loss: 0.5901 - val_accuracy: 0.6965 - val_loss: 0.5942
Epoch 3/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 15s 379ms/step - accuracy: 0.6442 - loss: 0.6326 - val_accuracy: 0.7114 - val_loss: 0.5330
Epoch 4/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 15s 374ms/step - accuracy: 0.7067 - loss: 0.5669 - val_accuracy: 0.6667 - val_loss: 0.5791
Epoch 5/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 15s 376ms/step - accuracy: 0.6961 - loss: 0.5706 - val_accuracy: 0.7264 - val_loss: 0.5269
Epoch 6/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 15s 375ms/step - accuracy: 0.7024 - loss: 0.5529 - val_accuracy: 0.7164 - val_loss: 0.5607
Epoch 7/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 15s 371ms/step - accuracy: 0.6858 - loss: 0.5613 - val_accuracy: 0.7114 - val_loss: 0.5657
Epoch 8/10
40/40 ━━━━━━━━━━━━━━━━━━━━ 15s 377ms/step - accuracy: 0.7001 - loss: 0.5653 - val_accu

In [68]:
np.mean(history.history['val_loss'])

0.5701210200786591

In [66]:
history.history['val_loss'][-5:]

[0.560728132724762,
 0.5656715035438538,
 0.5165605545043945,
 0.6440590023994446,
 0.5386602878570557]

#### Question 5 :
Mean of test loss for all the epochs for the model trained with augmentations is 0.57, close to the answer of 0.56

In [69]:
np.mean(history.history['val_accuracy'][-5:])

0.7084577083587646

#### Question 6 :
Average of test accuracy for the last 5 epochs (from 6 to 10) for the model trained with augmentations is 0.71, close to the answer of 0.71